CDAP-13924:Support filtering based on cdap-site.xml configuration

rohitsinha54 opened the pull request:
### Summary
- Filters plugin based on the requirement which has been provided by them and the configuration in `cdap-site.xml`
- Added a new property in `cdap-site.xml` called `plugin.exclude.with.requirements` which defines comma separated list of requirements which cannot be meet in the instance and any plugins requiring either one these should be excluded and be inaccessible to any user.
- Note: The change has unit tests at storage level. I am working on adding some unit test at higher level in handler to have end-to-end filtering tests. Opening PR now to get initial feedback on the change before weekend in case there are any major issues.

[Design](https://wiki.cask.co/display/CE/Improving+Plugins+User+Experience+Across+Different+Modes)
[Issue](https://issues.cask.co/browse/CDAP-13924)

Build: https://builds.cask.co/browse/CDAP-DUT6588-7

rohitsinha54 committed:
CDAP-13924:Support filtering based on cdap-site.xml configuration

albertshau commented on cdap-common/src/main/resources/cdap-default.xml:
```diff
@@ -625,6 +625,15 @@
+    <name>plugin.exclude.with.requirements</name>
+    <value></value>
+    <description>
+      Comma separated list of requirements which cannot be meet in the instance and any plugins requiring either one
```
Comma separated list of plugin requirements that cannot be met in the instance. Plugins can require any of these will be treated as if they don't exist. For example, if 'transactions' is given, any plugin that requires transactions will be treated as if they don't exist.

albertshau replied:
this change requires a change to the hash in the docs file.

rohitsinha54 replied:
Done.

albertshau commented on cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactStoreTest.java:
```diff
@@ -601,6 +607,58 @@ public void testGetAppClasses() throws Exception {
+    // plugins which have transaction as requirement should be excluded
+    Map<ArtifactDescriptor, Set<PluginClass>> actual = artifactStore.getPluginClasses(NamespaceId.DEFAULT, artifactId);
+    Assert.assertEquals(ImmutableMap.of(artifactInfo, ImmutableSet.of(includedPlugin, includedPlugin2)), actual);
+  }
```
Should also test the other store methods. For example if you list artifacts, the ArtifactDetail should not contain any information about the plugin that should be filtered. Same for getArtifact().

rohitsinha54 replied:
Yes, working on them right now. Will push a new commit for more test and separate one to address comments.

rohitsinha54 replied:
Done.

albertshau commented on cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactStoreTest.java:
```diff
@@ -84,7 +87,10 @@
-    artifactStore = AppFabricTestHelper.getInjector().getInstance(ArtifactStore.class);
+    CConfiguration cConf = CConfiguration.create();
+    // any plugin which requires transaction will be excluded
+    cConf.set(Constants.Plugin.EXCLUDE_WITH_REQUIREMENTS, Requirements.TRANSACTIONS);
```
Should test with multiple values in the exclude.

rohitsinha54 replied:
Good point.

rohitsinha54 replied:
Done.

albertshau reviewed (commented):
minor comments

rohitsinha54 commented:
@albertshau : Thanks a lot for super prompt review. Working on addressing comments.

anwar6953 commented:
What sort of UI work will be needed for https://issues.cask.co/browse/CDAP-14002 - if the exclusions are already specified at the cdap-site.xml level?
Will the UI filtering come into play if the cdap-site.xml is not excluding any?

rohitsinha54 commented:
@anwar6953 
The design discusses two levels of filtering. The UI changes is needed for second level: 

https://wiki.cask.co/display/CE/Improving+Plugins+User+Experience+Across+Different+Modes?focusedCommentId=8716406#ImprovingPluginsUserExperienceAcrossDifferentModes-PipelineDeveloperFiltering(Beyond5.1)

Although we are leaning towards moving it out of 5.1 as there are no definitie use case around it as of now but it is not decided yet.

rohitsinha54 commented:
@albertshau : Addressed all comments. Please have a look. The changes which needs to be review are the following two commits:

[Address Comments](https://github.com/caskdata/cdap/pull/10550/commits/b8a74eedfdea2fa4badb8c22e3996a06d94e5ed8): Addresses all the comments on the PR

[Add a handler level test](https://github.com/caskdata/cdap/pull/10550/commits/3a20042084688fc9b667da8676063c0aeaae3586): Adds a unit test at handler level for testing plugin exclusion.

albertshau reviewed (approved):
lgtm

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactStore.java:
```diff
@@ -189,9 +193,10 @@
 
   @Inject
-  ArtifactStore(DatasetFramework datasetFramework,
+  ArtifactStore(CConfiguration configuration, DatasetFramework datasetFramework,
```
Call the `Configuration` `cConf` to be consistent with other places.

rohitsinha54 replied:
Done

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactStore.java:
```diff
@@ -207,6 +212,9 @@
     );
     this.impersonator = impersonator;
+    this.excludedRequirements =
+      Arrays.stream(configuration.getTrimmedStrings(Constants.Plugin.EXCLUDE_WITH_REQUIREMENTS))
```
Use `new HashSet<>(cConf.getTrimmedStringCollection());`, which is simpler

rohitsinha54 replied:
Done.

chtyim commented on cdap-common/src/main/resources/cdap-default.xml:
```diff
@@ -625,6 +625,17 @@
   </property>
 
+  <property>
+    <name>plugin.exclude.with.requirements</name>
```
Why use exclusion rather that listing what are supported in the instance (inclusion)?

rohitsinha54 replied:
The exclusion approach looks clearer for the following reasons: 
1. If we take inclusion approach then we depend on the admin to specify all the possible requirements which are meet. So the default value has to include all the requirements meet.
2. The inclusion can be large list and hence someone will need to know them all and specify whereas exclusion will be a small list and hence easier to know and specify. If nothing is specified clearer to assume nothing is excluded.
3. With exclusion missing to specify a requirement will lead to plugin being displayed and then fail while running which can then be identified and added to exclusion list whereas for inclusion missing a requirement will lead to plugin not showing up at all and can lead to some confusion.

chtyim replied:
1. Shouldn't the admin should knows what are the supported in a given environment before making changes?
2. How large the list possibly be? We are talking about system abilities here, right?
3. What are the confusions?

rohitsinha54 replied:
It can be few requirements lets say 4-6 but we are relying on them to be specified to start with whereas we can easily work without them. Confusion of plugins not showing up. For example, if an admin in an instance does not say spark1 and/or spark2 and a plugin which is only compatible with spark1 is deployed will not be visible leading to the confusion of what happened by the the person who is deploying the jar (in most cases person deploying the jar is different than the admin and will not be aware of cdap-site setting)

What is the issue with having it as exclusion? Are there any demerits which favor inclusion more?

rohitsinha54 replied:
Had in person discussion and have decided to keep it exclusion for now and change the names. 
This commit has the changes discussed in person: https://github.com/caskdata/cdap/pull/10550/commits/e277b22c4993eadadcbdc1ab6bed61287642bb72

chtyim commented on cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactStoreTest.java:
```diff
@@ -601,6 +608,106 @@ public void testGetAppClasses() throws Exception {
+
+    // plugins which have transaction or spark as requirement should be excluded from plugins listed for the artifact
+    Map<ArtifactDescriptor, Set<PluginClass>> actual = artifactStore.getPluginClasses(NamespaceId.DEFAULT, artifactId);
+    ImmutableSet<PluginClass> expectedPlugins = ImmutableSet.of(includedPlugin1, includedPlugin2, includedPlugin3);
```
Use `Set` instead of `ImmutableSet` as the type.

rohitsinha54 replied:
Done.

chtyim commented on cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/http/handlers/ArtifactHttpHandlerTestBase.java:
```diff
@@ -0,0 +1,139 @@
+    EndpointStrategy endpointStrategy = new RandomEndpointStrategy(metadataHttpDiscovered);
+    Discoverable discoverable = endpointStrategy.pick(1, TimeUnit.SECONDS);
+    Assert.assertNotNull(discoverable);
+    String host = "127.0.0.1";
```
Why? The host should be coming from the `discoverable`?

rohitsinha54 replied:
Refactored to a base class code from existing code.
Fixed.

chtyim commented on cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/http/handlers/ArtifactHttpHandlerTestBase.java:
```diff
@@ -0,0 +1,139 @@
+    artifactRepository.clear(NamespaceId.SYSTEM);
+  }
+
+  void addSystemArtifacts() throws Exception {
```
Please add javadoc for methods that are not private.

rohitsinha54 replied:
Done.

chtyim commented on cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/http/handlers/ArtifactHttpHandlerTestBase.java:
```diff
@@ -0,0 +1,139 @@
+  }
+
+  // gets the contents from doing a get on the given url as the given type, or null if a 404 is returned
+  <T> T getResults(URL endpoint, Type type) throws IOException {
```
Missing `@Nullable`

rohitsinha54 replied:
Done.

chtyim commented on cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/http/handlers/PluginExclusionTest.java:
```diff
@@ -0,0 +1,79 @@
+  public static final ExternalResource RESOURCE = new ExternalResource() {
+    @Override
+    protected void before() {
+      // Set the system property to the excluded requirement value which will be set in cdap-site through the test base
```
Should reset/restore the property in `after()`

rohitsinha54 replied:
💯 Didn't know that. Thanks.

chtyim commented on cdap-api/src/main/java/co/cask/cdap/api/annotation/Requirements.java:
```diff
@@ -70,7 +70,7 @@
    * Defines transactional requirements
    */
-  String TRANSACTIONS = "transactions";
+  String TEPHRA_TX = "tephra-transactions";
```
Have the value as `tephraTx` as well.

rohitsinha54 replied:
Done.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactStore.java:
```diff
@@ -189,9 +192,10 @@
   private final DatasetFramework datasetFramework;
   private final Transactional transactional;
   private final Impersonator impersonator;
+  private final Set<String> excludedRequirements;
```
Call the field `requirementsBlackList` to be consistent with the config name

rohitsinha54 replied:
Done.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactStore.java:
```diff
@@ -1052,6 +1059,36 @@ private Scan scanAppClasses(NamespaceId namespace) {
+   * @param artifactMeta the artifact meta
+   * @return the {@link ArtifactMeta} containing only the plugins which are not excluded
+   */
+  private ArtifactMeta filterExcludedPlugins(ArtifactMeta artifactMeta) {
```
Call it `filterPlugins` as the method relies on the predicate `isAllow`, which can be more than exclude

rohitsinha54 replied:
Done.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactStore.java:
```diff
@@ -1052,6 +1059,36 @@ private Scan scanAppClasses(NamespaceId namespace) {
+   */
+  private boolean isAllowed(PluginClass pluginClass) {
+    // if a plugin has any requirement which is marked excluded in the config then the plugin should not be allowed
+    return Sets.intersection(pluginClass.getRequirements(), excludedRequirements).isEmpty();
```
Use `return pluginClass.getRequirements().stream().anyMatch(excludedRequirements::contains);` instead as there is no need to create a new set.

rohitsinha54 replied:
Done.

chtyim commented on cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/http/handlers/PluginExclusionTest.java:
```diff
@@ -0,0 +1,85 @@
+    @Override
+    protected void after() {
+      // clear the system property
+      System.clearProperty(Constants.REQUIREMENTS_BLACKLIST);
```
You should remember the old property in the `before()` and restore it here (or if the property did not exists in `before()`, then clear it here).

rohitsinha54 replied:
Thanks. Done.

chtyim commented on cdap-common/src/main/resources/cdap-default.xml:
```diff
@@ -625,6 +625,17 @@
+    <description>
+      Comma separated list of plugin requirements that cannot be met in the instance.
+      Plugins that require any of these will be treated as if they don't exist.
+      For example, if 'tephra-transactions' is given, any plugin that requires 'transactions' will be treated as
```
The example requirement doesn't match.

rohitsinha54 replied:
Done.

chtyim reviewed (approved):
LGTM

albertshau commented on cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactInspectorTest.java:
```diff
@@ -94,26 +94,26 @@ public void testGetPluginRequirements() {
 
     // test that duplicate requirements are only stored once and the beginning and ending white spaces are trimmed
-    Assert.assertEquals(ImmutableSet.of(Requirements.TEPHRA_TX, "duplicate"),
+    Assert.assertEquals(ImmutableSet.of(Requirements.TEPHRA_TX.toLowerCase(), "duplicate"),
```
may be better to make the constant itself all lowercase so that people don't see something different in their annotation and what comes back from the rest api.

rohitsinha54 replied:
Done.

rohitsinha54 commented:
Waiting on final build: https://builds.cask.co/browse/CDAP-DUT6588-9
